File: webapp/src/api/profiles.ts
import { apiClient } from "./client";

export interface Profile {
	id: string;
	url: string;
	profile_type: "company" | "personal";
	display_name: string;
	crawl_frequency_hours: number;
	is_active: boolean;
	last_crawled_at: string | null;
	created_at: string;
	updated_at: string;
}

export interface ProfileListResponse {
	items: Profile[];
	total: number;
	limit: number;
	offset: number;
}

export interface ProfileCreate {
	url: string;
	profile_type: "company" | "personal";
	display_name: string;
	crawl_frequency_hours?: number;
}

export interface ProfileUpdate {
	display_name?: string;
	crawl_frequency_hours?: number;
	is_active?: boolean;
}

export interface ProfileCrawlQueuedResponse {
	message: string;
	profile_id: string;
	profile_display_name: string;
}

export const profilesApi = {
	list: (params?: { limit?: number; offset?: number; is_active?: boolean }) => {
		const searchParams = new URLSearchParams();
		if (params?.limit) searchParams.set("limit", params.limit.toString());
		if (params?.offset) searchParams.set("offset", params.offset.toString());
		if (params?.is_active !== undefined)
			searchParams.set("is_active", params.is_active.toString());
		const query = searchParams.toString();
		return apiClient<ProfileListResponse>(
			`/profiles${query ? `?${query}` : ""}`,
		);
	},

	get: (id: string) => apiClient<Profile>(`/profiles/${id}`),

	create: (data: ProfileCreate) =>
		apiClient<Profile>("/profiles", {
			method: "POST",
			body: JSON.stringify(data),
		}),

	update: (id: string, data: ProfileUpdate) =>
		apiClient<Profile>(`/profiles/${id}`, {
			method: "PATCH",
			body: JSON.stringify(data),
		}),

	delete: (id: string) =>
		apiClient<void>(`/profiles/${id}`, {
			method: "DELETE",
		}),

	triggerCrawl: (id: string) =>
		apiClient<ProfileCrawlQueuedResponse>(`/profiles/${id}/crawl`, {
			method: "POST",
		}),
};

File: core/src/core/schemas/profiles.py
"""Profile schemas for API requests and responses."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, HttpUrl


class ProfileCreate(BaseModel):
    """Schema for creating a new monitored profile."""

    url: HttpUrl
    profile_type: Literal["company", "personal"]
    display_name: str = Field(min_length=1, max_length=200)
    crawl_frequency_hours: int = Field(default=24, ge=1, le=168)


class ProfileUpdate(BaseModel):
    """Schema for updating a monitored profile."""

    display_name: str | None = Field(default=None, min_length=1, max_length=200)
    crawl_frequency_hours: int | None = Field(default=None, ge=1, le=168)
    is_active: bool | None = None


class ProfileResponse(BaseModel):
    """Schema for profile response."""

    id: uuid.UUID
    url: str
    profile_type: str
    display_name: str
    crawl_frequency_hours: int
    is_active: bool
    last_crawled_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ProfileListResponse(BaseModel):
    """Schema for paginated profile list response."""

    items: list[ProfileResponse]
    total: int
    limit: int
    offset: int
